Security context - types, kubelet, admission

pweil- opened the pull request:
This PR is a subset of the original code in  https://github.com/GoogleCloudPlatform/kubernetes/pull/6287 and covers the following for SecurityContexts:
1.  New sub resource
2.  Kubelet support for setting flags based on SecurityContext Settings
3.  An always admit and always deny admission plugin
4.  Validation that defaults settings in the SecurityContext based on Container settings and vice versa

@erictune, some items I'd like to call out
1.  the SecurityContext is not inlined as recommended in https://github.com/GoogleCloudPlatform/kubernetes/pull/6287#discussion_r29000752 but I was unsure about the example for ResourceList that you mentioned.  If this implementation is incorrect I could use some guidance on where to look for the ResourceList defaulting
2.  I interpreted the "in the kubelet allow..." statements (https://github.com/GoogleCloudPlatform/kubernetes/pull/6287#issuecomment-95715700) as the kubelet always allows a SecurityContext if it makes it to the kubelet in a pod config.  Since I have not run a pod straight from a manifest I will confirm this on Monday in case I'm incorrect.
3.   I went ahead and used a provider as designed in https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/design/security_context.md#security-context-provider but it doesn't contain all the SecurityConstraint items as previously submitted.  If you'd still like to stay away from a formal provider and just have the logic in the kubelet I can move it.

I need to do some more testing on the functionality to make sure I carried it all over correctly but I wanted to get this under review while I did.

@erictune @smarterclayton @pmorie @liggitt @derekwaynecarr PTAL

History: https://github.com/GoogleCloudPlatform/kubernetes/pull/6287

erictune commented:
Suggest:
  - internal representation should only have `container.securityContext.privileged`, not `container.privileged`, and so on.
  - v1beta3, etc, have both
  - conversion from v1beta3 to internal fails on mismatch. 
  - conversion from internal to v1beta3 fills in both.
  - conversion and check happen in pkg/api/v1beta*/conversion.go
  - in a future version > v1beta3, we drop the `container.privileged` field.

@bgrant0607 check above statements if you have time.

pweil- commented:
10-4 - on it

bgrant0607 commented:
@erictune's proposal is correct. See container resources for an example where we used this approach.

erictune commented:
Conversion happens before admission control, and conversion sets corresponding values in SecurityContext if pod.privileged or pod.capabilities is set.  SecurityContextDeny disallows pods with any values set in SecurityContext.  So, are all pods with pod.privileged or pod.capabilities denied?  I think that is okay, but just wanted to clarify my understanding.

erictune commented on pkg/api/types.go:
```diff
@@ -1844,3 +1846,43 @@ type ComponentStatusList struct {
+// both the Container AND the SecurityContext will result in an error.
+type SecurityContext struct {
+	// Capabilities are the capabilities to add/drop when running the container
+	// DUPLICATE OF CONTAINER FIELD
```
replace `DUPLICATE OF CONTAINER FIELD` with `Must match Container.Privileged or be unset.`

erictune commented on pkg/api/types.go:
```diff
@@ -1844,3 +1846,43 @@ type ComponentStatusList struct {
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
+
+	// Run the container in privileged mode
+	// DUPLICATE OF CONTAINER FIELD
```
Same comment.

erictune commented on pkg/api/types.go:
```diff
@@ -1844,3 +1846,43 @@ type ComponentStatusList struct {
+
+// SELinuxOptions are the labels to be applied to the container
+type SELinuxOptions struct {
+	// User docker option --security-opt="label:user:USER"
```
@yifan-gu FYI this PR will adds docker knobs which I'm not sure if rkt supports yet.

yifan-gu replied:
@erictune Thanks for heads up! I will take a look.

jonboulle replied:
@erictune hrm, should this not be expressed in a less Docker-centric way, e.g. using the raw SELinux concepts?

smarterclayton replied:
Yes

> On Apr 28, 2015, at 7:42 PM, Jonathan Boulle <notifications@github.com> wrote:
> 
> In pkg/api/types.go:
> 
> > +	// Run the container in privileged mode
> > +	// DUPLICATE OF CONTAINER FIELD
> > +	Privileged *bool `json:"privileged,omitempty"`
> > +
> > +	// SELinuxOptions are the labels to be applied to the container
> > +	// and volumes
> > +	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
> > +
> > +	// RunAsUser is the UID to run the entrypoint of the container process.
> > +	// Docker option --user or -u
> > +	RunAsUser *int64 `json:"runAsUser,omitempty"`
> > +}
> > +
> > +// SELinuxOptions are the labels to be applied to the container
> > +type SELinuxOptions struct {
> > +	// User docker option --security-opt="label:user:USER"
> @erictune hrm, should this not be expressed in a less Docker-centric way, e.g. using the raw SELinux concepts?
> 
> —
> Reply to this email directly or view it on GitHub.
>

erictune replied:
@jonboulle yes, and suggestions on wording welcome.

erictune commented on pkg/api/v1beta1/types.go:
```diff
@@ -502,6 +502,8 @@ type Container struct {
 	// Optional: Capabilities for container.
 	Capabilities Capabilities `json:"capabilities,omitempty" description:"capabilities for container; cannot be updated"`
+	// Optional: SecurityContext defines the security options the pod should be run with
+	SecurityContext *SecurityContext `json:"securityContext,omitempty"`
```
add `description:"security options the pod should be run with"`

erictune commented on pkg/api/v1beta1/types.go:
```diff
@@ -1625,3 +1627,43 @@ type ComponentStatusList struct {
+type SecurityContext struct {
+	// Capabilities are the capabilities to add/drop when running the container
+	// DUPLICATE OF CONTAINER FIELD
+	Capabilities *Capabilities `json:"capabilities,omitempty"`
```
add `description:"capabilities to add/drop when running the container"`

erictune commented on pkg/api/v1beta1/types.go:
```diff
@@ -1625,3 +1627,43 @@ type ComponentStatusList struct {
+
+	// Run the container in privileged mode.  Pointer so we can tell if it was set or not
+	// DUPLICATE OF CONTAINER FIELD
+	Privileged *bool `json:"privileged,omitempty"`
```
add `description:"run container in privileged mode"`

erictune commented on pkg/api/v1beta1/types.go:
```diff
@@ -1625,3 +1627,43 @@ type ComponentStatusList struct {
+
+	// SELinuxOptions are the labels to be applied to the container
+	// and volumes
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
```
description in the struct tags.

erictune commented on pkg/api/v1beta1/types.go:
```diff
@@ -1625,3 +1627,43 @@ type ComponentStatusList struct {
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty"`
+
+	// RunAsUser is the UID to run the entrypoint of the container process.
+	// Docker option --user or -u
```
description in the struct tags.

erictune commented on pkg/api/v1beta1/types.go:
```diff
@@ -1625,3 +1627,43 @@ type ComponentStatusList struct {
+}
+
+// SELinuxOptions are the labels to be applied to the container
+type SELinuxOptions struct {
```
description in the struct tags everywhere

erictune commented on pkg/api/validation/validation.go:
```diff
@@ -1398,3 +1400,90 @@ func ValidateEndpointsUpdate(oldEndpoints, newEndpoints *api.Endpoints) errs.Val
+// 1.  if the container has no security context then create one and use all settings from the container
+// 2.  if the container has a security context but no capabilities defined then create one and use
+//		all capability settings from the container
+// 3.  if the security context has capabilities defined merge upwards or downwards by merging a non-zero
```
can we just reject if it has nonempty nonequal capabilities structs?

smarterclayton replied:
Probably - can't think of a user accidentally getting into this state.  Since validation is probably moving to external it seems reasonable.

> On Apr 28, 2015, at 7:37 PM, Eric Tune <notifications@github.com> wrote:
> 
> In pkg/api/validation/validation.go:
> 
> > +	allErrs := errs.ValidationErrorList{}
> > +	if seLinux.Disabled {
> > +		if len(seLinux.Level) > 0 || len(seLinux.Role) > 0 || len(seLinux.Type) > 0 || len(seLinux.User) > 0 {
> > +			allErrs = append(allErrs, errs.NewFieldInvalid("disabled", seLinux.Disabled, "cannot set disabled and labels"))
> > +		}
> > +	}
> > +	return allErrs
> > +}
> > +
> > +// mergeSecurityContextAndContainer performs a merge of the settings that exist on both the container
> > +// and the securitycontext.  This includes the Privileged field and Capabilities field
> > +// merge rules are:
> > +// 1.  if the container has no security context then create one and use all settings from the container
> > +// 2.  if the container has a security context but no capabilities defined then create one and use
> > +//		all capability settings from the container
> > +// 3.  if the security context has capabilities defined merge upwards or downwards by merging a non-zero
> can we just reject if it has nonempty nonequal capabilities structs?
> 
> —
> Reply to this email directly or view it on GitHub.
>

pweil- replied:
This section is mostly removed with the new implementation.  I am in the throes of a bad rebase and should have it fixed up soon.  In my current solution the defaulting takes place in the default settings for the v1beta* settings and there is no funky merge in validation.  Validation will be constrained to only ensuring correct values (non-negative user ids, not mixing selinux disabled with labels, etc) and the rejection for mismatched values takes place in the conversion process as recommended.

erictune commented on pkg/api/validation/validation.go:
```diff
@@ -1398,3 +1400,90 @@ func ValidateEndpointsUpdate(oldEndpoints, newEndpoints *api.Endpoints) errs.Val
+// mergeSecurityContextAndContainer performs a merge of the settings that exist on both the container
+// and the securitycontext.  This includes the Privileged field and Capabilities field
+// merge rules are:
+// 1.  if the container has no security context then create one and use all settings from the container
```
Should we just not create a context at all if privileged=false and capabilities to add and drop are empty lists?

erictune commented on plugin/pkg/admission/securitycontext/scadmit/admission.go:
```diff
@@ -0,0 +1,35 @@
+	"github.com/GoogleCloudPlatform/kubernetes/plugin/pkg/admission/admit"
+)
+
+func init() {
```
When I suggested implementing an always admission controller, I was thinking that you were also going to do some in the admission controller.  But, since there isn't any, I guess there is no need for this file.

erictune commented:
Almost there...

pweil- commented:
whew - git reflog saved me.  

@erictune the suggestion to move the check to the conversion really cleaned this up.  The defaulting happens in the default.go of each v1beta* now.  If an SC is already set then it won't be changed.  The conversion errors if the container fields don't match the SC fields that exist.  Otherwise a normal conversion occurs.  Validation is left to do normal validation-y checks.

Descriptions have been added to all fields in the feedback commit.

@yifan-gu, @jonboulle - I removed the direct references to the docker options in the comments for everything except disabled.  Perhaps disabled doesn't even need to be there, if you aren't specifying labels then maybe we just take no action.  If you have thoughts on how to express this better please let me know.

Thanks for the feedback so far.

pweil- commented:
> Conversion happens before admission control, and conversion sets corresponding values in SecurityContext if pod.privileged or pod.capabilities is set. SecurityContextDeny disallows pods with any values set in SecurityContext. So, are all pods with pod.privileged or pod.capabilities denied? I think that is okay, but just wanted to clarify my understanding.

As it stand right now, yes, it would.  It creates a conflict I didn't think of, the internal api is now completely based on SecurityContext but anything with SecurityContext is denied.   I'll fix that first thing tomorrow.  I'll allow a SecurityContext if it only has Privileged and Caps set.  Good catch.

erictune commented:
I'll take another look when you tell me you think you are done :smile:

pweil- commented:
some manual testing revealed the defaulting was simple...a little too simple.  I'm almost finished cleaning up the unit tests for the new code.  Should be done shortly

jonboulle commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,43 @@ type ComponentStatusList struct {
+	// SELinux level label.
+	Level string `json:"level,omitempty" description:"the level label to apply to the container"`
+
+	// Disabled docker option --security-opt="label:disable"
```
Right, seems to me this could just be omitted, and default to no label confinement?

jonboulle commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,43 @@ type ComponentStatusList struct {
+	SELinuxOptions *SELinuxOptions `json:"seLinuxOptions,omitempty" description:"options that control the SELinux labels applied"`
+
+	// RunAsUser is the UID to run the entrypoint of the container process.
+	// Docker option --user or -u
```
No need for this

pweil- commented:
Ok @erictune, @pmorie ready for another pass.  Tested with the following pod defs to ensure they were at least accepted/rejected properly:
1. using old container settings
1. no container settings and no security context
1. valid security context
1. invalid security context (caught by admission controller)
1. mismatched but valid security context (caught by converters)

@jonboulle - removed the missed reference to `--user` and removed the disabled option.

pmorie commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,39 @@ type ComponentStatusList struct {
+// SELinuxOptions are the labels to be applied to the container
+type SELinuxOptions struct {
+	// SELinux user label
+	User string `json:"user,omitempty" description:"the user label to apply to the container"`
```
Are these fields really optional?  Is it valid to, for example, only declare a type? Or a role and user, but no type?

pweil- replied:
yes, they have different knobs you can turn on and off in the docker settings.  It may not be a valid configuration but it is a possible configuration.

pmorie commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,39 @@ type ComponentStatusList struct {
+	RunAsUser *int64 `json:"runAsUser,omitempty" description:"the user id that runs the first process in the container"`
+}
+
+// SELinuxOptions are the labels to be applied to the container
```
Nit, this should be a sentence.

pmorie commented on pkg/api/v1/types.go:
```diff
@@ -639,6 +639,8 @@ type Container struct {
 	ImagePullPolicy PullPolicy `json:"imagePullPolicy" description:"image pull policy; one of PullAlways, PullNever, PullIfNotPresent; defaults to PullAlways if :latest tag is specified, or PullIfNotPresent otherwise; cannot be updated"`
 	// Optional: Capabilities for container.
```
Seems like there should be a note that this and privileged at this level are deprecated and to prefer the ones on security context.

pmorie replied:
Mark this as deprecated, same for privileged, in all API versions.

pmorie commented on pkg/api/v1beta2/conversion.go:
```diff
@@ -357,15 +358,21 @@ func init() {
+			if out.SecurityContext != nil && out.SecurityContext.Privileged != nil {
+				out.Privileged = *out.SecurityContext.Privileged
+			}
+			out.ImagePullPolicy = PullPolicy(in.ImagePullPolicy)
```
Seems like a bug since this field is dealt with on :361/:363 above.  Am I missing something?

pweil- replied:
Hmm, this may have been a bad rebase merge.  I'll investigate.

pweil- replied:
Yeah, bad merge.  It's handled in a conversion function now https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/api/v1beta2/conversion.go#L1174 .  I'll double check the other files.  Thanks for catching it

pmorie commented on pkg/api/validation/validation_test.go:
```diff
@@ -3180,3 +3180,89 @@ func TestValidateEndpoints(t *testing.T) {
+func TestValidateSecurityContext(t *testing.T) {
+	priv := false
+	var runAsUser int64 = 1
+	fullValidSC := func() *api.SecurityContext {
```
Style points for application of closures :clap:

pweil- replied:
:sunglasses:

pmorie commented on pkg/securitycontext/fake.go:
```diff
@@ -0,0 +1,45 @@
+)
+
+// ValidSecurityContextWithContainerDefaults creates a valid security context provider based on
+// empty container defaults.  Used for testing
```
Nit, add period after testing.

pmorie commented on pkg/securitycontext/provider.go:
```diff
@@ -0,0 +1,96 @@
+	docker "github.com/fsouza/go-dockerclient"
+)
+
+func NewSimpleSecurityContextProvider() SecurityContextProvider {
```
Nit, godoc

pmorie commented on pkg/securitycontext/provider_test.go:
```diff
@@ -0,0 +1,181 @@
+		expected []string
+	}{
+		{
+			"Empty val",
```
I think using field names would make this more readable (having not used them personally a number of times and spent a lot of time head-scratching when I came back to look at the code).

pmorie commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,39 @@ type ComponentStatusList struct {
+// both the Container AND the SecurityContext will result in an error.
+type SecurityContext struct {
+	// Capabilities are the capabilities to add/drop when running the container
+	// Must match Container.Capabilities or be unset.  Will be defaulted to Container.Capabilities if left unset
```
The 'must match Container.Capabilities' comment doesn't apply here since these are removed from the internal API.

pweil- replied:
This was specifically for https://github.com/GoogleCloudPlatform/kubernetes/pull/7343#discussion-diff-29297074.  They can still be specified in json so it'll help if folks are looking at a specific version definition

pmorie replied:
This line isn't relevant, remove

pmorie commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,39 @@ type ComponentStatusList struct {
+	Capabilities *Capabilities `json:"capabilities,omitempty" description:"the linux capabilites that should be added or removed"`
+
+	// Run the container in privileged mode
+	// Must match Container.Privileged or be unset.  Will be defaulted to Container.Privileged if left unset
```
Same thing here.

pmorie replied:
Remove

pmorie commented:
@pweil- The good news is, aside from some fairly minor nits, this LGTM and it looks like @erictune's comments have been addressed.  I'll leave it to eric to tag LGTM since he's the assignee.

The bad news is you need to rebase. :scream:

pweil- commented:
Ok, so the conversion unit test is broken on this commit.  test-cmd.sh and test-integration.sh are good.  Here's what happened

1.  We have a breaking change between internal <-> v1beta* 
1.  The converter will not generate functions because of the breaking change.  But not just for Containers, for anything that eventually has a Container
1.  The converter spit out errors and it appears those errors break the unit test when it does the programmatic comparisons

What I actually had to do to use the conversion utility

1.  First it won't compile since there is a breaking change in the types so remove all the functions and remove the init registration
1.  Run the conversion program and see what errors (shown below)
1.  Restore the old functions - they will be needed for overrides for non-generated functions
1.  Copy the new registration functions into init - now go through the diffs.  Some of the functions that were missing weren't reported in the errors by the converter utility (PodSpec, PodTemplateSpec, ReplicationControllerSpec, ReplicationControllerStatus)
1.  Copy from the old conversion functions into override functions
1.  Replace old conversion functions with new conversion functions
1.  Make my changes to the Container conversion functions

Simple as that :smile:  Suggestion box is now open, I'm not ruling out PEBKAC.

```
[pweil@localhost kubernetes]$ go run cmd/kube-conversion/conversion.go -v v1beta3 -f functions.txt -n init.txt
E0430 20:46:09.796432   22330 conversion.go:73] error while generating conversion functions for v1beta3.Pod: couldn't find a corresponding field Privileged in api.Container
E0430 20:46:09.796520   22330 conversion.go:73] error while generating conversion functions for v1beta3.ReplicationControllerList: couldn't find a corresponding field Privileged in api.Container
E0430 20:46:09.796551   22330 conversion.go:73] error while generating conversion functions for v1beta3.ReplicationController: couldn't find a corresponding field Privileged in api.Container
E0430 20:46:09.796574   22330 conversion.go:73] error while generating conversion functions for v1beta3.ListOptions: cannot convert types of different kinds: {LabelSelector  string json:"labelSelector" description:"a selector to restrict the list of returned objects by their labels; defaults to everything" 32 [1] false} {LabelSelector  labels.Selector  32 [1] false}
E0430 20:46:09.796600   22330 conversion.go:73] error while generating conversion functions for v1beta3.List: cannot convert types of different kinds: runtime.RawExtension runtime.Object
E0430 20:46:09.796623   22330 conversion.go:73] error while generating conversion functions for v1beta3.PodList: couldn't find a corresponding field Privileged in api.Container
E0430 20:46:09.796658   22330 conversion.go:73] error while generating conversion functions for v1beta3.PodTemplate: couldn't find a corresponding field Privileged in api.Container
E0430 20:46:09.796684   22330 conversion.go:73] error while generating conversion functions for v1beta3.PodTemplateList: couldn't find a corresponding field Privileged in api.Container
```

smarterclayton commented:
What do you mean by breaking change between internal and v1beta1?  That sounds more like there is a missing or invalid conversion.

> On Apr 30, 2015, at 9:53 PM, Paul <notifications@github.com> wrote:
> 
> Ok, so the conversion unit test is broken on this commit. test-cmd.sh and test-integration.sh are good. Here's what happened
> 
> We have a breaking change between internal <-> v1beta*
> The converter will not generate functions because of the breaking change. But not just for Containers, for anything that eventually has a Container
> The converter spit out errors and it appears those errors break the unit test when it does the programmatic comparisons
> What I actually had to do to use the conversion utility
> 
> First it won't compile since there is a breaking change in the types so remove all the functions and remove the init registration
> Run the conversion program and see what errors (shown below)
> Restore the old functions - they will be needed for overrides for non-generated functions
> Copy the new registration functions into init - now go through the diffs. Some of the functions that were missing weren't reported in the errors by the converter utility (PodSpec, PodTemplateSpec, ReplicationControllerSpec, ReplicationControllerStatus)
> Copy from the old conversion functions into override functions
> Replace old conversion functions with new conversion functions
> Make my changes to the Container conversion functions
> Simple as that  Suggestion box is now open, I'm not ruling out PEBKAC.
> 
> [pweil@localhost kubernetes]$ go run cmd/kube-conversion/conversion.go -v v1beta3 -f functions.txt -n init.txt
> E0430 20:46:09.796432   22330 conversion.go:73] error while generating conversion functions for v1beta3.Pod: couldn't find a corresponding field Privileged in api.Container
> E0430 20:46:09.796520   22330 conversion.go:73] error while generating conversion functions for v1beta3.ReplicationControllerList: couldn't find a corresponding field Privileged in api.Container
> E0430 20:46:09.796551   22330 conversion.go:73] error while generating conversion functions for v1beta3.ReplicationController: couldn't find a corresponding field Privileged in api.Container
> E0430 20:46:09.796574   22330 conversion.go:73] error while generating conversion functions for v1beta3.ListOptions: cannot convert types of different kinds: {LabelSelector  string json:"labelSelector" description:"a selector to restrict the list of returned objects by their labels; defaults to everything" 32 [1] false} {LabelSelector  labels.Selector  32 [1] false}
> E0430 20:46:09.796600   22330 conversion.go:73] error while generating conversion functions for v1beta3.List: cannot convert types of different kinds: runtime.RawExtension runtime.Object
> E0430 20:46:09.796623   22330 conversion.go:73] error while generating conversion functions for v1beta3.PodList: couldn't find a corresponding field Privileged in api.Container
> E0430 20:46:09.796658   22330 conversion.go:73] error while generating conversion functions for v1beta3.PodTemplate: couldn't find a corresponding field Privileged in api.Container
> E0430 20:46:09.796684   22330 conversion.go:73] error while generating conversion functions for v1beta3.PodTemplateList: couldn't find a corresponding field Privileged in api.Container
> —
> Reply to this email directly or view it on GitHub.
>

pweil- commented:
We have removed the Privileged and Capabilities field from internal but left them on v1beta* (https://github.com/GoogleCloudPlatform/kubernetes/pull/7343#issuecomment-96681075).  This did not play nicely with the conversion generator itself.

pweil- commented:
Digging in to the generator's unit test this morning.  @erictune if you want to take a look at the current state I think it's ready for your :hammer: 

/cc @pmorie - I've addressed your feedback in code or comments.  PTAL and apply your :hammer: as well.

pweil- commented:
Build is broken from the new rkt support.  Rebasing to check them out now.

erictune commented:
Remove WIP from description.
Looks like one unaddressed comment from pmorie about putting in comments about deprecation.
LGTM from an API and behavior standpoint, and code looks good with the caveat that I don't understand the conversion stuff very well.  I think pmorie knows it better.
At any rate, I'm out next week, so I'm assigning to pmorie for final review.

derekwaynecarr commented:
Please leave ResourceQuota as last admission controller.  I want all possible reasons for rejecting a pod to happen before quota usage is incremented.

pmorie commented on cluster/aws/config-default.sh:
```diff
@@ -72,4 +72,4 @@ DNS_DOMAIN="kubernetes.local"
 
 # Admission Controllers to invoke prior to persisting objects in cluster
-ADMISSION_CONTROL=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,ResourceQuota
+ADMISSION_CONTROL=NamespaceLifecycle,NamespaceAutoProvision,LimitRanger,ResourceQuota,SecurityContextDeny
```
Unaddressed comments from @derekwaynecarr re: ordering of admission control plugins

pmorie commented on pkg/api/types.go:
```diff
@@ -1865,3 +1863,39 @@ type ComponentStatusList struct {
 	Items []ComponentStatus `json:"items"`
 }
+
+// SecurityContext holds security configuration that will be applied to a container.  SecurityContext
```
I don't think this comment is correct for the context, since in the internal API these fields are not on Container.  I think we can probably leave this comment out (except for the first sentence).

pweil- replied:
Not sure I follow, the SecurityContext field is on the Container struct in the internal api and v1beta* apis.

pmorie commented on pkg/api/v1/conversion_test.go:
```diff
@@ -45,3 +45,63 @@ func TestNodeConversion(t *testing.T) {
+			}
+		}
+	}
+
```
Nit, newline

pmorie commented on pkg/api/v1beta1/defaults.go:
```diff
@@ -194,3 +195,44 @@ func defaultHostNetworkPorts(containers *[]Container) {
+		glog.V(4).Infof("downward merge of container.Privileged for container %s", container.Name)
+		container.SecurityContext.Privileged = &container.Privileged
+	} else {
+		//we don't have a good way to know if container.Privileged was set or just defaulted to false
```
Nit, missing spaces after `//` here, it looks weird since you use them throughout

pmorie commented on pkg/api/v1beta1/defaults_test.go:
```diff
@@ -340,3 +340,106 @@ func TestSetDefaultObjectFieldSelectorAPIVersion(t *testing.T) {
+	}
+}
+
+func isSecurityContextAndContainerEqual(c *current.Container) (bool, []string) {
```
Nit `is` -> `are`

pmorie commented on pkg/api/v1beta1/types.go:
```diff
@@ -521,6 +521,8 @@ type Container struct {
 	ImagePullPolicy PullPolicy `json:"imagePullPolicy" description:"image pull policy; one of PullAlways, PullNever, PullIfNotPresent; defaults to PullAlways if :latest tag is specified, or PullIfNotPresent otherwise; cannot be updated"`
 	// Optional: Capabilities for container.
```
Godoc note that this is deprecated.

pmorie commented on pkg/api/v1beta2/conversion_test.go:
```diff
@@ -564,3 +564,63 @@ func TestSecretVolumeSourceConversion(t *testing.T) {
+		c   *current.Container
+		err string
+	}{
+		//this use case must use true for the container and false for the sc.  Otherwise the defaulter
```
Nit, weird spacing after comment sigil

pmorie commented on pkg/api/v1beta2/defaults.go:
```diff
@@ -195,3 +196,44 @@ func defaultHostNetworkPorts(containers *[]Container) {
+		glog.V(4).Infof("downward merge of container.Privileged for container %s", container.Name)
+		container.SecurityContext.Privileged = &container.Privileged
+	} else {
+		//we don't have a good way to know if container.Privileged was set or just defaulted to false
```
Also needs spaces here

pmorie commented on pkg/api/v1beta2/defaults_test.go:
```diff
@@ -339,3 +339,106 @@ func TestSetDefaultObjectFieldSelectorAPIVersion(t *testing.T) {
+	}
+}
+
+func isSecurityContextAndContainerEqual(c *current.Container) (bool, []string) {
```
Nit `is` -> `are`

pmorie commented on pkg/api/v1beta3/conversion.go:
```diff
@@ -3227,3 +2746,474 @@ func init() {
 		panic(err)
 	}
 }
+
```
A lot of this diff is functions being relocated within the file without actual changes -- did the generator change and begin putting them in a different place?  I'd prefer to keep these functions (`convert_api_ReplicationController_To_v1beta3_ReplicationController` as an example) out of the diff if we can.

This is one of those things about code generation -- sometimes the generator changes and the whole file is a mess of diffs.  That said it looks like we can cut the diff back here to a great degree.

pweil- replied:
This was due to the change of removing a field from the internal api while retaining it in the v1beta* versions on the Container struct.  The conversion generator would not generate anything that had a field mismatch along with anything that contained the Container struct.  So yes, most of these are straight copies (retaining the naming convention as well).  The only actual meat in this file is the change to the container conversions.  The rest was a byproduct of the conversion utility no longer generating functions.

Important to note though, is that the functions must be moved.  The unit test for the auto generator uses the comments in the file to determine what to check to ensure that auto generated functions are not manually changed.  That is also why there is now the second call to Register the moved functions rather than leaving them in the RegisterAutoGenerated call.

pmorie replied:
Ok, got it.

pmorie commented on pkg/api/v1beta3/conversion.go:
```diff
@@ -3227,3 +2746,474 @@ func init() {
+	}
+	out.TerminationMessagePath = in.TerminationMessagePath
+	out.ImagePullPolicy = newer.PullPolicy(in.ImagePullPolicy)
+	if in.SecurityContext != nil {
```
TL;DR this if statement and subsequent `Convert` call consitute 50% of the real content of this diff.

pmorie commented on pkg/api/v1beta3/conversion.go:
```diff
@@ -3227,3 +2746,474 @@ func init() {
+	if err := s.Convert(&in.SecurityContext, &out.SecurityContext, 0); err != nil {
+		return err
+	}
+	// now that we've converted set the container field from security context
```
The other actual change in this diff.

pmorie commented on pkg/api/v1beta3/conversion_test.go:
```diff
@@ -45,3 +45,63 @@ func TestNodeConversion(t *testing.T) {
+		c   *current.Container
+		err string
+	}{
+		//this use case must use true for the container and false for the sc.  Otherwise the defaulter
```
Nit: spacing

pmorie commented on pkg/api/v1beta3/defaults.go:
```diff
@@ -156,3 +158,44 @@ func defaultHostNetworkPorts(containers *[]Container) {
+		glog.V(4).Infof("downward merge of container.Privileged for container %s", container.Name)
+		container.SecurityContext.Privileged = &container.Privileged
+	} else {
+		//we don't have a good way to know if container.Privileged was set or just defaulted to false
```
Spaces

pmorie commented on pkg/api/v1beta3/types.go:
```diff
@@ -634,11 +634,13 @@ type Container struct {
 	TerminationMessagePath string `json:"terminationMessagePath,omitempty" description:"path at which the file to which the container's termination message will be written is mounted into the container's filesystem; message written is intended to be brief final status, such as an assertion failure message; defaults to /dev/termination-log; cannot be updated"`
 	// Optional: Default to false.
-	Privileged bool `json:"privileged,omitempty" description:"whether or not the container is granted privileged status; defaults to false; cannot be updated"`
+	Privileged bool `json:"privileged,omitempty" description:"deprecated: whether or not the container is granted privileged status; defaults to false; cannot be updated  See SecurityContext."`
```
Description ought to be `whether or not the container is granted privileged status; defaults to false; deprecated; cannot be updated; see SecurityContext`

pmorie replied:
Similar wording for the other description, across all API versions

pmorie commented on pkg/kubelet/config/config_test.go:
```diff
@@ -62,7 +63,14 @@ func CreateValidPod(name, namespace, source string) *api.Pod {
+					Name:            "ctr",
+					Image:           "image",
+					ImagePullPolicy: "IfNotPresent",
+					SecurityContext: securitycontext.ValidSecurityContextWithContainerDefaults(),
```
I'm starting to wonder if we need an object mother method that decorates a pod spec with defaults.  That can be a follow up.

pweil- replied:
The defaulting takes place in the Container method.  The unit tests that were updated are hand crafting a pod to do their testing.  It might help to have a test utility that returns a valid container that can be manipulated if folks think it would be useful, the `securitycontext.ValidSecurityContextWithContainerDefaults()` method was a first stab.

pmorie replied:
I think this is okay for now.

pmorie commented on pkg/kubelet/rkt/rkt.go:
```diff
@@ -167,23 +168,29 @@ func rawValue(value string) *json.RawMessage {
 }
 
 // setIsolators overrides the isolators of the pod manifest if necessary.
+// TODO need an apply config in security context for rkt
```
Sweet, +1 for looking at this.

pmorie commented on pkg/securitycontext/util.go:
```diff
@@ -0,0 +1,39 @@
+package securitycontext
+
+import "github.com/GoogleCloudPlatform/kubernetes/pkg/api"
+
```
Godoc

pmorie commented on plugin/pkg/admission/securitycontext/scdeny/admission.go:
```diff
@@ -0,0 +1,65 @@
+type plugin struct {
+	client client.Interface
+}
+
```
Godoc

pmorie commented on plugin/pkg/admission/securitycontext/scdeny/admission.go:
```diff
@@ -0,0 +1,65 @@
+	}
+	for _, v := range pod.Spec.Containers {
+		if v.SecurityContext != nil {
+			if v.SecurityContext.SELinuxOptions != nil || v.SecurityContext.RunAsUser != nil {
```
Nit: you might want to break this into two different cases to be more specific with the error message, but I am fine doing that as a follow up.

pmorie commented:
Mostly there are nits that need to get fixed remaining.  I made a couple points about follow ups I would like.

pmorie commented:
@pweil- Have you done a GCE E2E run of this yet?

pmorie commented:
Also, once the nits are addressed, please:

1.  Squash these commits into:
    1.  API changes (ie, everything under `pkg/api`
    2.  All other changes
2.  Regenerate the swagger spec and make a separate commit for those changes.

Almost there!  I'm stoked about this going in.

pweil- commented:
@pmorie - some comments on some of the items you had feedback for and addressed the rest.  I have done a local e2e run with `--test="Pods,Docker Containers,Downward API,Networking,ReplicationController"`.  I'll do a GCE run as well.  If these look good then I'll move forward with the squash and regen swagger.  

I will also do one more pass at the generated conversion functions to make sure there was nothing introduced in master since there are so many changes in my file.  Thanks for the review

pmorie commented:
Okay. Looks like there is a separate thing to do to take care of swagger.
Don't worry about it for now. Will read your comments in detail later.
On Mon, May 4, 2015 at 8:00 PM Paul <notifications@github.com> wrote:

> @pmorie <https://github.com/pmorie> - some comments on some of the items
> you had feedback for and addressed the rest. I have done a local e2e run
> with --test="Pods,Docker Containers,Downward
> API,Networking,ReplicationController". I'll do a GCE run as well. If
> these look good then I'll move forward with the squash and regen swagger.
>
> I will also do one more pass at the generated conversion functions to make
> sure there was nothing introduced in master since there are so many changes
> in my file. Thanks for the review
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GoogleCloudPlatform/kubernetes/pull/7343#issuecomment-98887269>
> .
>

pmorie commented:
This LGTM; needs rebase.  Until #7779 goes in, we'll have to do the swagger changes as a separate PR.

Commit:
security context initial implementation - squash

pmorie commented:
Okay, even with #7779 merged, there is still an upstream go-restful bug that makes the validation tests break after the swagger spec is regenerated.  We need that fixed before we can regen the swagger spec after this.

pmorie commented:
Actually, I think I was wrong.  I ran the script locally and it produced the correct output.

Commit:
update swagger

pmorie commented:
LGTM.  I will merge on green.

pmorie commented on api/swagger-spec/v1beta1.json:
```diff
@@ -7623,7 +7627,8 @@
      },
      "terminationGracePeriodSeconds": {
-      "$ref": "int64",
+      "type": "integer",
```
@bgrant0607 I'm not sure why this wasn't changed in #7779, but it looks like the pointers are being handled correctly now.

pmorie commented:
Merged

zmerlynn commented:
This was reverted, and was the cause of #7805. See there for more context.

pmorie commented:
For the record, the revert was itself reverted.

smarterclayton commented:
It was verted.

> On May 5, 2015, at 8:12 PM, Paul Morie <notifications@github.com> wrote:
> 
> For the record, the revert was itself reverted.
> 
> —
> Reply to this email directly or view it on GitHub.
>
